text/0000-derive-enum-default.md

```diff
@@ -0,0 +1,363 @@
+- Feature Name: `derive_enum_default`
+- Start Date: 2021-04-07
+- RFC PR: TODO
+- Rust Issue: TODO
+
+# Summary
+[summary]: #summary
+
+An attribute `#[default]`, usable on `enum` variants, is also introduced, thereby allowing enums to
+work with `#[derive(Default)]`.
+
+```rust
+#[derive(Default)]
+enum Foo {
+    #[default]
+    Alpha(u8),
+    Beta,
+    Gamma,
+}
+
+assert_eq!(Foo::default(), Foo::Alpha(0));
+```
+
+The `#[default]` attribute may not be used on a variant that is also declared `#[non_exhaustive]`.
+
+# Motivation
+[motivation]: #motivation
+
+## `#[derive(Default)]` in more cases
+
+Currently, `#[derive(Default)]` is not usable for `enum`s. To rectify this situation, a `#[default]`
+attribute is introduced that can be attached to variants. This allows you to use
+`#[derive(Default)]` on enums wherefore you can now write:
+
+```rust
+// from time
+#[derive(Default)]
+enum Padding {
+    Space,
+    Zero,
+    #[default]
+    None,
+}
+```
+
+## Clearer documentation and more local reasoning
+
+Providing good defaults when such exist is part of any good design that makes a physical tool, UI
+design, or even data-type more ergonomic and easily usable. However, that does not mean that the
+defaults provided can just be ignored and that they need not be understood. This is especially the
+case when you are moving away from said defaults and need to understand what they were. Furthermore,
+it is not too uncommon to see authors writing in the documentation of a data-type that a certain
+value is the default.
+
+All in all, the defaults of a data-type are therefore important properties. By encoding the defaults
+right where the data-type is defined gains can be made in terms of readability particularly with
+regard to. the ease of skimming through code. In particular, it is easier to see what the default
+variant is if you can directly look at the `rustdoc` page and read:
+
+```rust
+#[derive(Default)]
+enum Foo {
+    #[default]
+    Bar {
+        alpha: u8,
+    },
+    Baz {
+        beta: u16,
+        gamma: bool,
+    }
+}
+```
+
+This way, you do not need to open up the code of the `Default` implementation to see what the
+default variant is.
```

Comment:
This section feels like it carries a lot of the weight of the Motivation section, if only by word count, and thus makes it seem as if the primary benefits of this RFC are to enhance rustdoc. However, this RFC does not include anything, as far as I can tell, that _explicitly_ binds rustdoc to displaying the desired result, which is notable since rustdoc output is, as far as I am aware, not something that is considered so stable that it is guaranteed to show the exact declaration specified here without developer intervention.

If such a strong amount of the RFC hinges on rustdoc output, why could this segment of the RFC not be a direct functional improvement to rustdoc where it extracts the information for the Default value from the impl of Default for a type and displays it? This would also improve the experience for existing users without any changes required in user code, merely rebuilding the docs.

I bring this up because at the time of this writing, there is a significant visual difference between stable and nightly rustdoc output, e.g. struct declarations being hidden behind a button which may or may not be explicitly expanded or not. Compare:
https://doc.rust-lang.org/stable/std/vec/struct.Vec.html
https://doc.rust-lang.org/beta/std/vec/struct.Vec.html
https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html

As far as I know, the way this is displayed has changed a few times before and may change again in the future. So if nothing else, these concerns and alternatives should probably be discussed regarding why we aren't doing it instead, or doing them first. It may not address the entire concern of this RFC, and it may be possible to do both, but prioritization is not an insignificant thing.

Reply:
Even if it's collapsed by default (note that this is only for enums, not structs, and these are currently expanded on stable), it would still be an improvement imo to have `#[default]` shown. Right now the only way know the default is to open up the source code for the trait impl (aside from the author explicitly stating it, of course). I don't think it would make sense for rustdoc to have to introspect the implementation in any situation. The primary reason for this RFC is to permit easier `Default` impls for enums.

cc @GuillaumeGomez on whether we could promise this going forward?

Reply:
Thanks @workingjubilee for bringing it up!

Generally, rustdoc follows the implementation. However, the output of rustdoc is not stable and it's generally up to the one(s) implementing a feature to then either implement its counterpart in rustdoc or contact the rustdoc team so it can be implemented by us.

Now, there were a lot of changes in rustdoc output in the last months, on the type definitions too. If the RFC is about to add support for `#[default]` on variants, we can always talk about how to implement it and how it should be displayed in rustdoc afterwards. It's never easy, and generally take quite some time for small details. In this case, would it be enough to simply display `#[default]` on the enum declaration? Because if there are too many variants, they'll be collapsed by default and the information won't be accessible right away to the users' eyes (like in https://doc.rust-lang.org/nightly/std/io/enum.ErrorKind.html). So should we also display this information in the variant list just below (https://doc.rust-lang.org/nightly/std/io/enum.ErrorKind.html#variants) and if so how? Maybe even make it "stand out" a bit in the sidebar too? Like I said, lot of small details.

So in short: if this is a new feature for the rust compiler first, then we can talk about rustdoc afterwards. However, if this is a rustdoc-first feature, then it'll need more discussion I think.

Reply:
Gotcha. I'll just cut back on the depth here, as this certainly isn't a rustdoc-first feature (that's just an aside, really). I can move it to "future possibilities".

Reply:
Sounds good to me! Please contact the rustdoc team when implemented then so we can take a look at it. :)

Reply:
>  I don't think it would make sense for rustdoc to have to introspect the implementation in any situation.

My understanding is a significant amount of how rustdoc works meaningfully amounts to introspection of Rust code anyways, as it operates by simply driving the compiler, so I don't think "introspection of implementation" is fundamentally something rustdoc shouldn't do. It may be a meaningfully more complicated extension, of course, which is a completely valid reason as to why to take one path over another.

Reply:
Moved to "future possibilities" in the latest push.